src/secure/signed.rs

```
@@ -0,0 +1,206 @@
+    /// assert_ne!(jar.find("signed").unwrap().value(), "text");
+    /// assert!(jar.find("signed").unwrap().value().contains("text"));
+    ///
+    /// // They can be verified through the child jar.
+    /// assert_eq!(jar.signed(&key).find("signed").unwrap().value(), "text");
+    ///
+    /// // A tampered with cookie does not validate but still exists.
+    /// let mut cookie = jar.find("signed").unwrap().clone();
+    /// jar.add(Cookie::new("signed", cookie.value().to_string() + "!"));
+    /// assert!(jar.signed(&key).find("signed").is_none());
+    /// assert!(jar.find("signed").is_some());
+    /// ```
+    fn signed(&'a mut self, key: &'k [u8]) -> SignedJar<'a, 'k>;
+}
+
+impl<'a, 'k> Signed<'a, 'k> for CookieJar {
+    fn signed(&'a mut self, key: &'k [u8]) -> SignedJar<'a, 'k> {
+        if key.len() != KEY_LEN {
+            panic!("bad key length: expected {} bytes, found {}", KEY_LEN, key.len());
+        }
+
+        SignedJar { parent: self, key: key }
+    }
+}
+
+/// A child cookie jar that authenticates its cookies.
+///
+/// A _signed_ child jar signs all the cookies added to it and verifies cookies
+/// retrieved from it. Any cookies stored in a `SignedJar` are assured integrity
+/// and authenticity. In other words, clients cannot tamper with the contents of
+/// a cookie nor can they fabricate cookie values, but the data is visible in
+/// plaintext.
+///
+/// This type is only available when the `secure` feature is enabled.
+pub struct SignedJar<'a, 'k> {
+    parent: &'a mut CookieJar,
+    key: &'k [u8]
+}
+
+impl<'a, 'k> SignedJar<'a, 'k> {
+    /// Given a signed value `str` where the signature is prepended to `value`,
+    /// verifies the signed value and returns it. If there's a problem, returns
+    /// an `Err` with a string describing the issue.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `value.len()` < BASE64_DIGEST_LEN.
+    fn verify(&self, cookie_value: &str) -> Result<String, &'static str> {
```

Why panic when the value is too short, instead of returning `Err`?